### ISSUE TITLE
To ensure minimal port collisions, the output should be uniformly distributed

### ISSUE THREAD

**Issue by Finnventor** [HenrikBengtsson/port4me#71]
From my understanding, the LCG gives uniform output only when given random data as an input.
However, the seeds aren't uniformly distributed: Most seeds will be lowercase ASCII, so nearly every byte is `010XXXXX`, and vowels are more common. I think the question becomes, is our hash function correctly converting that into uniformly distributed seeds?

As a quick check, here's a histogram of port4me's output with the first username dataset I found:
https://github.com/danielmiessler/SecLists/blob/master/Usernames/xato-net-10-million-usernames.txt
![port4me_distribution](https://github.com/HenrikBengtsson/port4me/assets/50817575/0eb0acf5-2efc-41de-b75d-6b0d61ee40ca)

```
import matplotlib.pyplot as plt
from port4me import port4me

port_numbers = []

with open('xato-net-10-million-usernames.txt') as f:
    i = 0
    for line in f:
        port_numbers.append(port4me(user=line.strip()))
        if i % 100000 == 0:
            print(i, end='\r')
        i += 1
print(i)

plt.hist(port_numbers, bins='auto')
plt.xlabel('Port number')
plt.ylabel('Frequency')

plt.savefig('port4me_distribution.png')
plt.show()
```

(I recommend short-circuiting `is_port_free` and `get_env_ports` to save time when running this test.)

_Originally posted by @Finnventor in https://github.com/HenrikBengtsson/port4me/issues/60#issuecomment-1890835808_

**Comment by HenrikBengtsson** [comment]
> From my understanding, the LCG gives uniform output only when given random data as an input.

It's actually uniform regardless of the initial seed $s_0$.  For instance, if the LCG algorithm starts out with the initial seed $s_0 = 0$, the LCG algorithm will hit all values in $\[0, 65535\]$ [exactly once](https://github.com/HenrikBengtsson/port4me/blob/065183625198454821e6af641ae65fb8b82ffd76/r/tests/lcg-uniformity.R#L48), _before_ it hits an already hit value a second time.  I've added unit tests for $s_0 = 0$ some days ago, and I just now added tests for a few more initial seeds.

EDIT +24h: It's quite clear that the LCG has to visit all numbers exactly once before being able to hit an already visited number. If it wouldn't, that would mean those unvisited numbers won't be hit later either, because it's a fully deterministic sequence that is generated. As soon as it hits an already visited number, it'll basically start over again.

**Comment by HenrikBengtsson** [comment]
So, assuming my conclusions about LCG being uniform is correct, the point you're raising here does not have to worry about the random algorithm per se, but rather about clashes of the initial seed $s_0$ between usernames.

(watching a movie here, but I'll check back later)

**Comment by HenrikBengtsson** [comment]
> I recommend short-circuiting is_port_free and get_env_ports to save time when running this test.

So, effectively, this means that your graph is basically showing the distribution of initial seeds $\\{s_0\\}$ across usernames, correct?  And, $s_0$ is calculated via `uint_hash(username)`.

**Comment by Finnventor** [comment]
Yes, it's the distribution of `port4me(user=username)`, which is the distribution of `lcg(uint_hash(username + ','))` except ports less than 1024 make the LCG select a new value, as usual.

**Comment by HenrikBengtsson** [comment]
Thanks for confirming.  

The distribution of `lcg(uint_hash(username + ','))` looks quite uniform to me. One could add a formal measure to it, but since this is not rocket science, eying it is good enough.
